File: api/models/squad.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Index, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class Squad(Base):
    """A persistent pool of players an organiser draws on for their games.

    A squad is deliberately allowed to be larger than a game's `max_players`:
    every member is invited to each match and slots are filled
    first-come-first-served by whoever accepts.
    """

    __tablename__ = "squads"

    id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    organiser_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    __table_args__ = (Index("ix_squads_organiser_id", "organiser_id"),)


class SquadMember(Base):
    """One player's membership of a squad.

    `user_id` is nullable on purpose: an organiser can add a friend who has no
    BeTheFifth account yet. Those rows sit in `invited` status carrying only a
    display name and an email, and are linked to a real user by
    `SquadService.link_pending_members` when that person registers.
    """

    __tablename__ = "squad_members"

    id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    squad_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("squads.id"), nullable=False
    )
    user_id: Mapped[uuid.UUID | None] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    display_name: Mapped[str] = mapped_column(String(100), nullable=False)
    invite_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    invite_phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    status: Mapped[str] = mapped_column(
        String(20), default="invited", nullable=False
    )  # invited (awaiting sign-up), active (linked to a real user)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    __table_args__ = (
        # NULLs compare as distinct, so these only constrain rows that actually
        # carry a user_id / invite_email — multiple pending invites coexist fine.
        UniqueConstraint("squad_id", "user_id", name="uq_squad_member_user"),
        UniqueConstraint("squad_id", "invite_email", name="uq_squad_member_email"),
        Index("ix_squad_members_squad_id", "squad_id"),
        Index("ix_squad_members_user_id", "user_id"),
        Index("ix_squad_members_invite_email", "invite_email"),
    )

File: app/lib/models/squad.dart
/// One player's membership of a squad.
///
/// [userId] is null for someone the organiser invited by email who hasn't
/// registered yet — the backend links them to a real account on sign-up.
class SquadMember {
  final String id;
  final String squadId;
  final String? userId;
  final String displayName;
  final String? inviteEmail;
  final String? invitePhone;
  final String status; // invited, active
  final DateTime createdAt;

  const SquadMember({
    required this.id,
    required this.squadId,
    this.userId,
    required this.displayName,
    this.inviteEmail,
    this.invitePhone,
    this.status = 'invited',
    required this.createdAt,
  });

  /// True while this member has no BeTheFifth account yet. They can't be
  /// invited to individual matches until they register.
  bool get isPending => userId == null;

  factory SquadMember.fromJson(Map<String, dynamic> json) {
    return SquadMember(
      id: json['id'] as String,
      squadId: json['squad_id'] as String,
      userId: json['user_id'] as String?,
      displayName: json['display_name'] as String,
      inviteEmail: json['invite_email'] as String?,
      invitePhone: json['invite_phone'] as String?,
      status: json['status'] as String? ?? 'invited',
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// An organiser's persistent pool of regular players.
///
/// Deliberately allowed to be larger than a game's capacity — everyone is
/// invited to each match and slots go first-come-first-served.
class Squad {
  final String id;
  final String organiserId;
  final String name;
  final DateTime createdAt;

  /// Only populated by the detail endpoint. The list endpoint omits members
  /// and sends counts instead, so never derive counts from this — use
  /// [activeCount] / [pendingCount].
  final List<SquadMember> members;

  final int activeCount;
  final int pendingCount;

  const Squad({
    required this.id,
    required this.organiserId,
    required this.name,
    required this.createdAt,
    this.members = const [],
    this.activeCount = 0,
    this.pendingCount = 0,
  });

  factory Squad.fromJson(Map<String, dynamic> json) {
    final rawMembers = json['members'] as List<dynamic>?;
    final members = rawMembers == null
        ? const <SquadMember>[]
        : rawMembers
            .map((e) => SquadMember.fromJson(e as Map<String, dynamic>))
            .toList();

    return Squad(
      id: json['id'] as String,
      organiserId: json['organiser_id'] as String,
      name: json['name'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      members: members,
      activeCount: json['active_member_count'] as int? ??
          members.where((m) => !m.isPending).length,
      pendingCount: json['pending_member_count'] as int? ??
          members.where((m) => m.isPending).length,
    );
  }
}
